flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java

```diff
@@ -203,8 +221,11 @@ private void updateAvailability() {
 
     private void updateAvailability(
             DataInputStatus status, StreamOneInputProcessor<?> input, int inputIdx) {
-        if (status == DataInputStatus.MORE_AVAILABLE
-                || (status != DataInputStatus.END_OF_INPUT && input.isApproximatelyAvailable())) {
+        if (status == DataInputStatus.END_OF_DATA) {
+            inputSelectionHandler.setDataFinishedOnInput(inputIdx);
+        } else if (status == DataInputStatus.END_OF_INPUT) {
+            inputSelectionHandler.setEndOfPartition(inputIdx);
+        } else if (status == DataInputStatus.MORE_AVAILABLE || input.isApproximatelyAvailable()) {
```

Comment:
You are potentially changing order of evaluation of the conditions, pushing down `MORE_AVAILALBE` hot path.

Again, I'm not certain of that, but I wouldn't personally risk that (given instabilities of our two/multi input benchmarks). Performance regression of couple % would be virtually impossible to track down.

Reply:
That's fine. `input.isApproximatelyAvailable()`  shouldn't be on the hot path.